src/v/raft/types.cc

```diff
@@ -249,6 +251,162 @@ std::ostream& operator<<(std::ostream& o, const install_snapshot_reply& r) {
     return o;
 }
 
+ss::future<> heartbeat_request::serde_async_write(iobuf& dst) {
+    vassert(!heartbeats.empty(), "cannot serialize empty heartbeats request");
+
+    struct sorter_fn {
+        constexpr bool operator()(
+          const raft::heartbeat_metadata& lhs,
+          const raft::heartbeat_metadata& rhs) const {
+            return lhs.meta.commit_index < rhs.meta.commit_index;
+        }
+    };
+
+    iobuf out;
+    auto& request = *this;
+
+    std::sort(
+      request.heartbeats.begin(), request.heartbeats.end(), sorter_fn{});
+
+    co_await ss::coroutine::maybe_yield();
+
+    internal::hbeat_soa encodee(request.heartbeats.size());
+    // target physical node id is always the same it differs only by
+    // revision
+
+    const size_t size = request.heartbeats.size();
+    for (size_t i = 0; i < size; ++i) {
+        const auto& m = request.heartbeats[i].meta;
+        const raft::vnode node = request.heartbeats[i].node_id;
+        const raft::vnode target_node = request.heartbeats[i].target_node_id;
+        vassert(m.group() >= 0, "Negative raft group detected. {}", m.group);
+        encodee.groups[i] = m.group;
+        encodee.commit_indices[i] = std::max(model::offset(-1), m.commit_index);
+        encodee.terms[i] = std::max(model::term_id(-1), m.term);
+        encodee.prev_log_indices[i] = std::max(
+          model::offset(-1), m.prev_log_index);
+        encodee.prev_log_terms[i] = std::max(
+          model::term_id(-1), m.prev_log_term);
+        encodee.last_visible_indices[i] = std::max(
+          model::offset(-1), m.last_visible_index);
+        encodee.revisions[i] = std::max(
+          model::revision_id(-1), node.revision());
+        encodee.target_revisions[i] = std::max(
+          model::revision_id(-1), target_node.revision());
+    }
+    // important to release this memory after this function
+    // request.meta = {}; // release memory
+
+    using serde::write;
+
+    // physical node ids are the same for all requests
+    write(out, request.heartbeats.front().node_id.id());
+    write(out, request.heartbeats.front().target_node_id.id());
+    write(out, static_cast<uint32_t>(size));
+
+    co_await ss::coroutine::maybe_yield();
```

Comment:
maybe it would be better to put this inside the first `for` loop, this way we will be able to break the long running loop into smaller synchronous chunks.

Reply:
i will do this.